psi4/src/psi4/libfock/DirectJK.cc

```diff
@@ -908,4 +912,1372 @@ void DirectJK::build_JK(std::vector<std::shared_ptr<TwoBodyAOInt>>& ints, std::v
+            size += primary_->shell(task_shells[P2]).nfunction();
+        }
+        max_task = (max_task >= size ? max_task : size);
+    }
+
+    if (debug_) {
+        outfile->Printf("  ==> DirectJK: Task Blocking <==\n\n");
+        for (size_t task = 0; task < ntask; task++) {
+            outfile->Printf("  Task: %3d, Task Start: %4d, Task End: %4d\n", task, task_starts[task],
+                            task_starts[task + 1]);
+            for (int P2 = task_starts[task]; P2 < task_starts[task + 1]; P2++) {
+                int P = task_shells[P2];
+                int size = primary_->shell(P).nfunction();
+                int off = primary_->shell(P).function_index();
+                int off2 = task_offsets[P2];
+                outfile->Printf("    Index %4d, Shell: %4d, Size: %4d, Offset: %4d, Offset2: %4d\n", P2, P, size, off,
+                                off2);
+            }
+        }
+        outfile->Printf("\n");
+    }
+
+    // => Significant Task Pairs (PQ|-style <= //
+
+    std::vector<std::pair<int, int>> task_pairs;
+    for (size_t Ptask = 0; Ptask < ntask; Ptask++) {
+        for (size_t Qtask = 0; Qtask < ntask; Qtask++) {
+            if (Qtask > Ptask) continue;
+            bool found = false;
+            for (int P2 = task_starts[Ptask]; P2 < task_starts[Ptask + 1]; P2++) {
+                for (int Q2 = task_starts[Qtask]; Q2 < task_starts[Qtask + 1]; Q2++) {
+                    int P = task_shells[P2];
+                    int Q = task_shells[Q2];
+                    if (ints[0]->shell_pair_significant(P, Q)) {
+                        found = true;
+                        task_pairs.push_back(std::pair<int, int>(Ptask, Qtask));
+                        break;
+                    }
+                }
+                if (found) break;
+            }
+        }
+    }
+
+    size_t ntask_pair = task_pairs.size();
+    size_t ntask_pair2 = ntask_pair * ntask_pair;
+
+    // => Intermediate Buffers <= //
+
+    std::vector<std::vector<std::shared_ptr<Matrix> > > JKT;
+    for (int thread = 0; thread < nthread; thread++) {
+        std::vector<std::shared_ptr<Matrix> > JK2;
+        for (size_t ind = 0; ind < D.size(); ind++) {
+            JK2.push_back(std::make_shared<Matrix>("JKT (J only)", 2 * max_task, max_task));
+        }
+        JKT.push_back(JK2);
+    }
+
+    // => Benchmarks <= //
+
+    size_t computed_shells = 0L;
+
+// ==> Master Task Loop <== //
+
+#pragma omp parallel for num_threads(nthread) schedule(dynamic) reduction(+ : computed_shells)
+    for (size_t task = 0L; task < ntask_pair2; task++) {
+        size_t task1 = task / ntask_pair;
+        size_t task2 = task % ntask_pair;
+
+        int Ptask = task_pairs[task1].first;
+        int Qtask = task_pairs[task1].second;
+        int Rtask = task_pairs[task2].first;
+        int Stask = task_pairs[task2].second;
+
+        // GOTCHA! Thought this should be RStask > PQtask, but
```

Comment:
I don't understand this comment. Can you translate? Again, applies to all the places this appears.